### PR TITLE
Improve parser

### PR DESCRIPTION
**Pull request by DaniPopes**
This PR does 3 main things to the `solang-parser` crate:
1. adds documentation to all items
2. improves some logic in the parsing functions, like less allocations and some minor refactoring
3. most importantly, adds parse tree helpers (moving existing implementations):
  - `fmt`: implements `std::fmt::Display` for all parse tree types
  - `loc`, expanding on the `CodeLocation` traits:
    - implementations for all possible parse tree types
    - generic implementations for `Vec`, `Option` etc
    - renames the `loc` method to `loc_opt` on `OptionalCodeLocation` to avoid name collision
    - previously there were some standalone (not with the traits) methods on structs, these were moved to the trait implementations
  - `ord`: implements `std::cmp::{Ord, PartialOrd}`
  - helper methods on parse tree types; you can view them by looking for the `impl` blocks in `pt`

The massive diff in the lexer tests is just whitespace and renaming the fully qualified `collect::<Vec<...>>` to just `Vec<_>`

Some of these changes have been upstreamed from [foundry-rs/foundry](https://github.com/foundry-rs/foundry)'s `forge-doc` and `forge-fmt`.

BREAKING CHANGES:
- `pt::OptionalCodeLocation::loc` renamed to `loc_opt`
- `pt::Level::to_string` renamed to `as_str`
- `lexer::Spanned<Token, Loc, Error>` split into `Spanned<'a>` and `Result<'a, T = Spanned<'a>, E = LexicalError>`
- `pt::UsingList`'s `Error` variant changed from tuple to unit (no more `()` at the end)

### CODE REVIEW

**Comment by seanyoung** [comment]
The new files need:
```
// SPDX-License-Identifier: Apache-2.0
```
on the first line, else CI will fail. Also note the clippy errors.

File: solang-parser/src/doccomment.rs [hunk 1]
```diff
@@ -1,27 +1,65 @@
 // SPDX-License-Identifier: Apache-2.0
 
+//! Solidity parsed doc comments.
+//!
+//! See also the Solidity documentation on [natspec].
+//!
+//! [natspec]: https://docs.soliditylang.org/en/latest/natspec-format.html
```
Comment (xermicus) [inline solang-parser/src/doccomment.rs:7]:
```suggestion
//! See also the Solidity documentation on [natspec](https://docs.soliditylang.org/en/latest/natspec-format.html).
```

Markdown links are supported here.

File: solang-parser/src/helpers/fmt.rs [hunk 2]
```diff
@@ -0,0 +1,2518 @@
...
+        }
+    }
+}
+
+impl Display for pt::ContractTy {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        f.write_str(self.as_str())
+    }
+}
+impl pt::ContractTy {
```
Comment (xermicus) [inline solang-parser/src/helpers/fmt.rs:413]:
```suggestion
}

impl pt::ContractTy {
```

File: solang-parser/src/helpers/fmt.rs [hunk 3]
```diff
@@ -0,0 +1,2518 @@
...
+            }
+            Self::Default(_, block) => {
+                f.write_str("default ")?;
+                block.fmt(f)
+            }
+        }
+    }
+}
+
+#[inline]
```
Comment (xermicus) [inline solang-parser/src/helpers/fmt.rs:1211]: Can you elaborate here why we need this? It is is almost always unnecessary and deserves  a comment that proves this is needed here.

File: solang-parser/src/helpers/fmt.rs [hunk 4]
```diff
@@ -0,0 +1,2518 @@
...
+fn fmt_parameter_list(list: &pt::ParameterList, f: &mut Formatter<'_>) -> Result {
+    let iter = list.iter().flat_map(|(_, param)| param);
+    write_separated_iter(iter, f, ", ")
+}
+
+#[inline]
```
Comment (xermicus) [inline solang-parser/src/helpers/fmt.rs:1217]: Same

File: solang-parser/src/helpers/fmt.rs [hunk 5]
```diff
@@ -0,0 +1,2518 @@
...
+        //     }
+        // };
+        // ( assembly { $($t:tt)* } ) => {
+        //     pt::Statement::Assembly {
+        //         loc: loc!(),
+        //         dialect: None,
+        //         flags: None,
+        //         block: pt::YulBlock,
+        //     }
+        // };
```
Comment (xermicus) [inline solang-parser/src/helpers/fmt.rs:1400]:
```suggestion
```

**Review by xermicus — COMMENTED** [review]
Thanks for your contribution!

File: solang-parser/src/doccomment.rs (hunk 1 above)
Comment (DaniPopes) [reply solang-parser/src/doccomment.rs:7]: That is also a valid markdown link, and makes reading easier IMO.

File: solang-parser/src/helpers/fmt.rs (hunk 3 above)
Comment (DaniPopes) [reply solang-parser/src/helpers/fmt.rs:1211]: very simple functions that only call another function, I don't see why not to add this; same for all the helper functions that do simple matching or just a field access

File: solang-parser/src/helpers/fmt.rs (hunk 2 above)
Comment (DaniPopes) [reply solang-parser/src/helpers/fmt.rs:413]: Display `impl` is grouped with its corresponding struct `impl`, but I can change that if you want to

Comment (xermicus) [reply solang-parser/src/helpers/fmt.rs:413]: Well I thought this slipped by. Anyways it's a nitpick, I'll leave it up to you

File: solang-parser/src/doccomment.rs (hunk 1 above)
Comment (xermicus) [reply solang-parser/src/doccomment.rs:7]: Very nice, I didn't know that

File: solang-parser/src/helpers/fmt.rs (hunk 3 above)
Comment (xermicus) [reply solang-parser/src/helpers/fmt.rs:1211]: This function looks a lot like it should be getting inlined by the compiler automatically, especially considering it's private. If I see this `#[inline]` here I'd assume that the compiler doesn't inline this function when it should
and it is used in a hot path somewhere where removing it introduces significant performance penalties.

If these assumptions are correct, it would be nice to leave a comment explaining that, otherwise the `#[inline]` is probably unnecessary.

Comment (seanyoung) [reply solang-parser/src/helpers/fmt.rs:1211]: The compiler knows perfectly well when things should be inlined or not. It does not sound there is a solid reason for the `#[inline]` attribute.

File: solang-parser/src/pt.rs [hunk 6]
```diff
@@ -1,1003 +1,1688 @@
...
+    /// The list of values.
+    ///
+    /// This field contains `None` only if an error occurred during parsing.
     pub values: Vec<Option<Identifier>>,
 }
 
+/// A variable attribute.
 #[derive(Debug, PartialEq, Eq, Clone)]
 #[cfg_attr(feature = "pt-serde", derive(Serialize, Deserialize))]
+#[repr(u8)] // for cmp; order of variants is important
```
Comment (LucasSte) [inline solang-parser/src/pt.rs:833]: Why is this necessary?

File: solang-parser/src/pt.rs [hunk 7]
```diff
@@ -1,1003 +1,1688 @@
...
+    /// `payable`
+    Payable(Loc),
 }
 
+/// Function visibility.
+///
+/// Deprecated for [FunctionTy] other than `Function`.
 #[derive(Debug, PartialEq, Eq, Clone)]
 #[cfg_attr(feature = "pt-serde", derive(Serialize, Deserialize))]
+#[repr(u8)] // for cmp; order of variants is important
```
Comment (LucasSte) [inline solang-parser/src/pt.rs:1290]: Why do you need the representation to be `u8`?

Comment (DaniPopes) [reply solang-parser/src/pt.rs:1290]: it's needed to add the Ord implementations by using only the discriminant, see the cmp file. But I realize it might be better off just deriving the implementations

File: solang-parser/src/pt.rs [hunk 8]
```diff
@@ -1,1003 +1,1688 @@
...
 pub enum UserDefinedOperator {
+    /// `&`
     BitwiseAnd,
+    /// `~`
+    ///
+    /// AKA BitwiseNot
     Complement,
+    /// `-`
+    ///
+    /// Note that this is the same as `Subtract`, and that it is currently not being parsed.
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply solang-parser/src/pt.rs:587]: There is no way to distinguish them without semantic analysis

[the remark this answers is not in the document]
Comment (seanyoung) [reply solang-parser/src/pt.rs:587]: sema uses it.

This is not related to this PR.

File: solang-parser/src/pt.rs (hunk 6 above)
Comment (DaniPopes) [reply solang-parser/src/pt.rs:833]: The problem with using the derive macro is that it imposes the trait bound on all of the fields, which will also be checked during the call.
For example, deriving `PartialOrd` for `FunctionAttribute` transitively requires the trait to be implemented for `Mutability`, `IdentifierPath`, and `Base`; we can achieve this by implementing it manually, which goes against the whole point of deriving in the first place, or deriving it for these types too, which will eventually require deriving for almost the entire `pt` module.

By doing it manually we can make ordering depend only on the discriminant, regardless of the variants' value/fields; I think is fine for the few types (attributes, visibility) that have a recommended order in the Solidity standards.

File: solang-parser/src/helpers/ord.rs [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+                fn cmp(&self, other: &Self) -> Ordering {
+                    Ord::cmp(self.as_discriminant(), other.as_discriminant())
+                }
+            }
+        )+
+    };
+}
+
+// SAFETY: every type must be `repr(u8)` for this to be safe, see comments in macro implementation.
+impl_with_cast!(pt::Visibility, pt::VariableAttribute, pt::FunctionAttribute);
```
Comment (seanyoung) [inline solang-parser/src/helpers/ord.rs:43]: I do totally get why you've done it this way, but I'd like to avoid `unsafe {}`. What do you think of the following?

```suggestion
impl pt::Visibility {
    fn solidity_order(&self) -> u8 {
        match self {
            pt::Visibility::External(..) => 0,
            pt::Visibility::Internal(..) => 1,
            pt::Visibility::Public(..) => 2,
            pt::Visibility::Private(..) => 3,
        }
    }
}

impl PartialOrd for pt::Visibility {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.solidity_order().cmp(&other.solidity_order()))
    }
}

impl Ord for pt::Visibility {
    fn cmp(&self, other: &Self) -> Ordering {
        self.solidity_order().cmp(&other.solidity_order())
    }
}

impl pt::VariableAttribute {
    fn solidity_order(&self) -> u8 {
        match self {
            pt::VariableAttribute::Visibility(..) => 0,
            pt::VariableAttribute::Constant(..) => 1,
            pt::VariableAttribute::Immutable(..) => 2,
            pt::VariableAttribute::Override(..) => 3,
        }
    }
}

impl PartialOrd for pt::VariableAttribute {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.solidity_order().cmp(&other.solidity_order()))
    }
}

impl Ord for pt::VariableAttribute {
    fn cmp(&self, other: &Self) -> Ordering {
        self.solidity_order().cmp(&other.solidity_order())
    }
}

impl pt::FunctionAttribute {
    fn solidity_order(&self) -> u8 {
        match self {
            pt::FunctionAttribute::Visibility(..) => 0,
            pt::FunctionAttribute::Mutability(..) => 1,
            pt::FunctionAttribute::Virtual(..) => 2,
            pt::FunctionAttribute::Immutable(..) => 3,
            pt::FunctionAttribute::Override(..) => 4,
            pt::FunctionAttribute::BaseOrModifier(..) => 5,
            pt::FunctionAttribute::Error(..) => 6,
        }
    }
}

impl PartialOrd for pt::FunctionAttribute {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.solidity_order().cmp(&other.solidity_order()))
    }
}

impl Ord for pt::FunctionAttribute {
    fn cmp(&self, other: &Self) -> Ordering {
        self.solidity_order().cmp(&other.solidity_order())
    }
}
```

**Review by seanyoung — COMMENTED** [review]
There is a lot of good stuff in this - very large - pull request, it is great, thank you very much!

So I only have one comment, please us let know what you think. Thanks for some quality code changes. :pray:

File: solang-parser/src/helpers/ord.rs (hunk 9 above)
Comment (DaniPopes) [reply solang-parser/src/helpers/ord.rs:43]: That's how I originally implemented it, though I didn't feel like it was really idiomatic. Seeing as the standard Discriminant doesn't implement Ord, I then tried this method which is recommended in the Rust docs. I believe it's also easier to maintain and has no overhead due to it being just a pointer cast.
But if you would like to avoid unsafe at all costs then I can change this, up to you.

File: solang-parser/src/helpers/fmt.rs (hunk 3 above)
Comment (DaniPopes) [reply solang-parser/src/helpers/fmt.rs:1211]: > The compiler knows perfectly well when things should be inlined or not. It does not sound there is a solid reason for the `#[inline]` attribute.

I don't believe that is the case. You are probably right in saying that since these functions are private then the compiler is more likely to inline them, but giving extra hints that they should be doesn't hurt, especially because they are private and that they should probably be inlined regardless.

> If these assumptions are correct, it would be nice to leave a comment explaining that (ideally pointing to some benchmarks or tests), otherwise the `#[inline]` is probably unnecessary.

I don't have benchmarks nor do I intend to write any simply to justify a mere compiler hint that may or may not sway performance and/or compilation speed by less than a single percentage point. Since there are almost no downsides to adding these hints I would rather just leave them as is.

This being said, if you still want me to remove the `#[inline]` hints on these private functions, I will do so.

Comment (DaniPopes) [reply solang-parser/src/helpers/fmt.rs:1211]: > They are a maintenance burden, especially without any comment providing context.

Given:
- `fn a<T>(x: T, y, z) { ... }`
- `#[inline] fn b(x: &[u8]) { a(x, <default_y>, <default_z>) }`

Since `b` is just a simple shortcut for a special case of `a`, is the reason to add `#[inline]` not obvious, or at least deductible, without requiring 2 paragraphs of comments detailing why this should be the case?

I could've just not written `b` and instead write manual calls to `a` everywhere, but since I noticed a common pattern for the same type I wrote `b` to make easier and cleaner, and to reduce any possible performance impact I add a hint that it should be the same as the manual `a` calls.

But alas, if you so insist I'll add a comment linking to this discussion.

File: solang-parser/src/helpers/fmt.rs [hunk 10]
```diff
@@ -0,0 +1,2501 @@
...
+            }
+            Self::Default(_, block) => {
+                f.write_str("default ")?;
+                block.fmt(f)
+            }
+        }
+    }
+}
+
+#[inline]
```
Comment (xermicus) [inline solang-parser/src/helpers/fmt.rs:1211]: It is obvious why it should be inlined. What is not obvious is why the compiler doesn't do it automatically or why we think it needs this hint (we _don't_ know) and so why the need for this attribute arises.

I left the comments as a suggestion, WDYT? :upside_down_face: 

```suggestion
// This function is private so it should be inlined by the compiler anyways. 
// We provided this as an extra hint regardless because we don't expect compile time penalties or other negative impacts from it.
#[inline]
```

File: solang-parser/src/helpers/fmt.rs [hunk 11]
```diff
@@ -0,0 +1,2501 @@
...
+fn fmt_parameter_list(list: &pt::ParameterList, f: &mut Formatter<'_>) -> Result {
+    let iter = list.iter().flat_map(|(_, param)| param);
+    write_separated_iter(iter, f, ", ")
+}
+
+#[inline]
```
Comment (xermicus) [inline solang-parser/src/helpers/fmt.rs:1217]:
```suggestion
// This function is private so it should be inlined by the compiler anyways. 
// We provided this as an extra hint regardless because we don't expect compile time penalties or other negative impacts from it.
#[inline]
```

File: solang-parser/src/helpers/fmt.rs (hunk 3 above)
Comment (xermicus) [reply solang-parser/src/helpers/fmt.rs:1211]: https://github.com/hyperledger/solang/pull/1237/files#r1151621676

File: solang-parser/src/helpers/ord.rs (hunk 9 above)
Comment (seanyoung) [reply solang-parser/src/helpers/ord.rs:43]: It's a matter of taste - let's go with your version. Thank you for the discussion.

**Review by seanyoung — APPROVED** [review]

File: solang-parser/src/helpers/fmt.rs (hunk 3 above)
Comment (LucasSte) [reply solang-parser/src/helpers/fmt.rs:1215]: @DaniPopes Inlining functions is not guaranteed to improve performance. It might also make the code slower:
https://isocpp.org/wiki/faq/inline-functions#inline-and-perf

**Review by LucasSte — APPROVED** [review]

**Comment by seanyoung** [comment]
@DaniPopes thank you very much, great stuff! :pray:
